src/psyclone/psyir/nodes/dynamic_omp_task_directive.py

```diff
@@ -0,0 +1,1998 @@
+                        start = ref.ancestor(StructureReference)
+                        sub_ref = StructureReference(start.symbol)
+                        sub_ref.addchild(start.member.copy())
+                        array_member = sub_ref.walk(Member)[-1]
+                        num_child = len(array_member.children)
+                        array_member.pop_all_children()
+                        for _ in range(num_child):
+                            array_member.addchild(one.copy())
+                    else:
+                        arrayref = ref.parent.parent
+                        sub_ref = Reference(arrayref.symbol)
+                    # Create a full range object for this symbol and
+                    # children
+                    lbound = BinaryOperation.create(
+                        BinaryOperation.Operator.LBOUND,
+                        sub_ref.copy(),
+                        one.copy(),
+                    )
+                    ubound = BinaryOperation.create(
+                        BinaryOperation.Operator.UBOUND,
+                        sub_ref.copy(),
+                        one.copy(),
+                    )
+                    full_range = Range.create(lbound, ubound)
+                    index_list.append(full_range)
+                else:
+                    # We have a private constant, written to inside
+                    # our region, so we can't do anything better than
+                    # a full range I think (since we don't know what
+                    # the value is/how it changes.
+                    # Return a full range (:)
+                    dim = len(index_list)
+                    one = Literal(str(dim + 1), INTEGER_TYPE)
+                    arrayref = ref.parent.parent
+                    lbound = BinaryOperation.create(
+                        BinaryOperation.Operator.LBOUND,
+                        Reference(arrayref.symbol),
+                        one.copy(),
+                    )
+                    ubound = BinaryOperation.create(
+                        BinaryOperation.Operator.UBOUND,
+                        Reference(arrayref.symbol),
+                        one.copy(),
+                    )
+                    full_range = Range.create(lbound, ubound)
+                    index_list.append(full_range)
+            else:
+                if ref not in firstprivate_list:
+                    firstprivate_list.append(ref.copy())
+                if ref.symbol in self._parent_loop_vars:
+                    # Non-proxy access to a parent loop variable.
+                    # In this case we have to do similar to when accessing a
+                    # proxy loop variable.
+
+                    # Find index of parent loop var
+                    ind = self._parent_loop_vars.index(ref.symbol)
+                    parent_loop = self._parent_loops[ind]
+                    # We have a Literal step value, and a Literal in
```

Comment:
factored most of this code out (with the lambda change you also suggested) into a new function `_create_binops_from_step_and_divisors`. It feels a little awkward to describe/document but since its an internal-only function hopefully its not too major an issue.

Reply:
Unit test added too.

Reply:
Great, thanks.